### PR TITLE
Update Docker Personality to support Docker API 1.25 [full ci]

### LINKED ISSUE #3415: Update docker personality to accept API versions > 1.23
As a user of VIC, I want to use the latest version of the docker CLI and use it with the latest version of VIC without getting a version mismatch error like the following,
```
Error response from daemon: client is newer than server (client API version: 1.24, server API version: 1.23)
```
As a user of VIC, I should also not have to set DOCKER_API_VERSION environment variable just to use VIC.

# One option:
We need to update the docker personality server to support the upcoming API version 1.25.

1. Update vendored docker sources to 1.13
2. Remove vendored engine-api as it is now deprecated and moved into the main docker sources
3. Update the docker personality's backend code
4. Update the docker personality's main.go code

### PR DESCRIPTION
**Pull request by sflxn**
1. Updated vendored sources to support Docker 1.13
2. Updated all the backend interfaces within
   vic/lib/apiservers/backends
3. Added hollowed out swarm.go and backends/exec/SwarmBackend.go
4. Updated main.go
5. Added robot scripts for docker secret, checkpoint, deploy, node,
   plugin, service, stack, swarm.

resolves #3415

### CODE REVIEW

File: lib/apiservers/engine/backends/cache/image_cache.go [hunk 1]
```diff
@@ -24,8 +24,8 @@ import (
 
 	log "github.com/Sirupsen/logrus"
 
-	"github.com/docker/distribution/digest"
-	derr "github.com/docker/docker/errors"
+	// "github.com/docker/distribution/digest"
```
Comment (vburenin) [inline lib/apiservers/engine/backends/cache/image_cache.go]: remove it?

File: lib/apiservers/engine/backends/container_proxy.go [hunk 2]
```diff
@@ -84,7 +85,7 @@ type VicContainerProxy interface {
 	CommitContainerHandle(handle, containerID string, waitTime int32) error
 	StreamContainerLogs(name string, out io.Writer, started chan struct{}, showTimestamps bool, followLogs bool, since int64, tailLines int64) error
 
-	Stop(vc *viccontainer.VicContainer, name string, seconds int, unbound bool) error
+	Stop(vc *viccontainer.VicContainer, name string, seconds *int, unbound bool) error
```
Comment (vburenin) [inline lib/apiservers/engine/backends/container_proxy.go]: why pointer?

File: lib/apiservers/engine/backends/container_proxy.go [hunk 3]
```diff
@@ -711,11 +712,20 @@ func (c *ContainerProxy) AttachStreams(ctx context.Context, vc *viccontainer.Vic
 	plClient, transport := c.createNewAttachClientWithTimeouts(attachConnectTimeout, 0, attachAttemptTimeout)
 	defer transport.Close()
 
+	keys := []byte{}
```
Comment (vburenin) [inline lib/apiservers/engine/backends/container_proxy.go]: in this case it cases memory allocation. Try to use "var keys []byte"

File: lib/apiservers/engine/backends/endpoint.go [hunk 4]
```diff
@@ -129,3 +129,9 @@ func (e *endpoint) Address() *net.IPNet {
 func (e *endpoint) AddressIPv6() *net.IPNet {
 	return nil
 }
+
+// DisableService removes a managed contianer's endpoints from the load balancer
```
Comment (vburenin) [inline lib/apiservers/engine/backends/endpoint.go]: This comment is definitely invalid. Could you fix it?

File: lib/apiservers/engine/backends/executor/SwarmBackend.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
+package executor
+
+import (
+	"io"
+	"time"
+
```
Comment (vburenin) [inline lib/apiservers/engine/backends/executor/SwarmBackend.go]: remove empty line here.

File: lib/apiservers/engine/backends/container_proxy.go (hunk 2 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/container_proxy.go]: This is Docker's change.

**Comment by vburenin** [comment]
Let me know when you need review.

**Comment by sflxn** [comment]
@vburenin Sure thing.  I still need to get networking fully working and write robot scripts to verify we don't croak on swarm, checkpoint, build, and plugin calls.

File: lib/apiservers/engine/backends/checkpoint.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
```
Comment (vburenin) [inline lib/apiservers/engine/backends/checkpoint.go]: it is 2017 outside, at least it was so yesterday ;)

File: lib/apiservers/engine/backends/executor/SwarmBackend.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	networktypes "github.com/docker/libnetwork/types"
+	"github.com/docker/swarmkit/agent/exec"
+	"golang.org/x/net/context"
+)
+
+type SwarmBackend struct {
+}
+
+func (b SwarmBackend) CreateManagedNetwork(clustertypes.NetworkCreateRequest) error {
+	return nil
```
Comment (hickeng) [inline lib/apiservers/engine/backends/executor/SwarmBackend.go]: Am I correct in assuming that we have to make this a noop instead an "unsupported" error?

File: lib/apiservers/engine/backends/cache/image_cache.go [hunk 8]
```diff
@@ -183,10 +182,10 @@ func (ic *ICache) Get(idOrRef string) (*metadata.ImageConfig, error) {
 	return copyImageConfig(config), nil
 }
 
-func (ic *ICache) getImageByDigest(digest digest.Digest) *metadata.ImageConfig {
+func (ic *ICache) getImageByDigest(digest string) *metadata.ImageConfig {
```
Comment (hickeng) [inline lib/apiservers/engine/backends/cache/image_cache.go]: is this change to signature needed - from just this commit it looks like the original code cast it to string where needed and it's got to be a string type anyway given the comparison on line 167

File: lib/apiservers/engine/backends/executor/SwarmBackend.go (hunk 7 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/executor/SwarmBackend.go]: Docker called it a "backend," but it's not a REST router backend.  The actual REST router backend is in lib/apiservers/engine/backends/swarm.go.

File: lib/apiservers/engine/backends/checkpoint.go (hunk 6 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/checkpoint.go]: good catch :)

File: lib/apiservers/engine/backends/cache/image_cache.go (hunk 8 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/cache/image_cache.go]: I cannot remember why I did this.  I remember it was to get around a build issue, but that was before 1.13 officially got released.  With the updated vendored files, the old signature works so I reverted it back.

**Comment by hmahmood** [comment]
Does this fix #3712?

**Comment by hmahmood** [comment]
nvm ... I guess you left the networking part out.

**Comment by sflxn** [comment]
@hmahmood Yeah, I left it out.  Before fully hooking up the cluster management code, I wanted to see if minimally hooking it up would pass our integration tests so we can avoid unnecessary work.

File: lib/apiservers/engine/backends/plugins.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backends
+
+import (
+    "fmt"
```
Comment (jzt) [inline lib/apiservers/engine/backends/plugins.go]: minor: indent is short

File: lib/apiservers/engine/backends/sandbox.go [hunk 10]
```diff
@@ -20,6 +20,7 @@ import (
 	"github.com/docker/docker/pkg/stringid"
 	"github.com/docker/libnetwork"
 	"github.com/docker/libnetwork/types"
+	// "k8s.io/minikube/pkg/minikube/notify"
```
Comment (jzt) [inline lib/apiservers/engine/backends/sandbox.go]: Not sure if you left this on purpose, just bringing it to your attention :)

File: tests/test-cases/Group1-Docker-Commands/1-32-Docker-Plugin.robot [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+Suite Setup  Install VIC Appliance To Test Server  certs=${false}
+Suite Teardown  Cleanup VIC Appliance On Test Server
+
+*** Test Cases ***
+Docker plugin install
+    ${rc}  ${output}=  Run And Return Rc And Output  docker1.13 %{VCH-PARAMS} plugin install vieux/sshfs
+    Should Be Equal As Integers  ${rc}  1
+    Should Contain  ${output}  does not yet support plugins
+
+#Docker plugin create
```
Comment (jzt) [inline tests/test-cases/Group1-Docker-Commands/1-32-Docker-Plugin.robot]: Do we still need/want this test? Could we maybe pass & warn or something?

File: cmd/docker/main.go [hunk 12]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
+// Copyright 2016 VMware, Inc. All Rights Reserved.
```
Comment (jzt) [inline cmd/docker/main.go]: Looks like you accidentally removed the copyright

File: lib/apiservers/engine/backends/cache/image_cache.go [hunk 13]
```diff
@@ -166,6 +166,7 @@ func (ic *ICache) Get(idOrRef string) (*metadata.ImageConfig, error) {
 	var config *metadata.ImageConfig
 	if imgDigest != "" {
 		config = ic.getImageByDigest(imgDigest)
+		// config = ic.getImageByDigest(string(imgDigest))
```
Comment (jzt) [inline lib/apiservers/engine/backends/cache/image_cache.go]: Commented code

File: lib/apiservers/engine/backends/image.go [hunk 14]
```diff
@@ -27,10 +27,12 @@ import (
 
 	"github.com/docker/docker/pkg/streamformatter"
 	"github.com/docker/docker/reference"
-	"github.com/docker/engine-api/types"
-	"github.com/docker/engine-api/types/container"
-	"github.com/docker/engine-api/types/filters"
-	"github.com/docker/engine-api/types/registry"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/backend"
+	// "github.com/docker/docker/api/types/container"
```
Comment (jzt) [inline lib/apiservers/engine/backends/image.go]: commented import

File: lib/apiservers/engine/backends/cache/image_cache.go [hunk 15]
```diff
@@ -184,6 +185,7 @@ func (ic *ICache) Get(idOrRef string) (*metadata.ImageConfig, error) {
 }
 
 func (ic *ICache) getImageByDigest(digest digest.Digest) *metadata.ImageConfig {
+// func (ic *ICache) getImageByDigest(digest string) *metadata.ImageConfig {
```
Comment (jzt) [inline lib/apiservers/engine/backends/cache/image_cache.go]: commented code

**Comment by jzt** [comment]
LGTM

File: lib/apiservers/engine/backends/plugins.go (hunk 9 above)
Comment (sflxn) [inline lib/apiservers/engine/backends/plugins.go]: fixed

File: tests/test-cases/Group1-Docker-Commands/1-32-Docker-Plugin.robot (hunk 11 above)
Comment (sflxn) [inline tests/test-cases/Group1-Docker-Commands/1-32-Docker-Plugin.robot]: Fixed that test.

**Pull request merged**